### PR TITLE
[core] out of band serialization exception

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR

**This PR is backward compatible**

- Introduce an env var to raise an exception when there's out of band seriailzation of object ref
- Improve error message on out of band serialization issue. There are 2 types of issues. 1. cloudpikcle.dumps(ref). 2. implicit capture. See below for more details.
- Update an anti-pattern doc. 

cloudpickle.dumps error message
```
E                       ray.exceptions.RayTaskError(OufOfBandRefSerializationException): ray::f() (pid=61703, ip=127.0.0.1)
E                         File "/Users/sangcho/work/ray/python/ray/tests/test_serialization.py", line 751, in f
E                           cloudpickle.dumps(ray.put(1))
E                         File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1479, in dumps
E                           cp.dump(obj)
E                         File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1245, in dump
E                           return super().dump(obj)
E                       ray.exceptions.OufOfBandRefSerializationException: It is not allowed to serialize ray.ObjectRef 00ef45ccd0112571ffffffffffffffffffffffff0100000002e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
E                       Callsite: Disabled. Set RAY_record_ref_creation_sites=1
```

implicit capture error 
```
2024-09-06 17:47:28,806	INFO worker.py:1777 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
Traceback (most recent call last):
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 73, in pickle_dumps
    return pickle.dumps(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 152, in object_ref_reducer
    self.add_contained_object_ref(obj, obj.call_site())
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 221, in add_contained_object_ref
    raise ray.exceptions.OufOfBandRefSerializationException(
ray.exceptions.OufOfBandRefSerializationException: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sangcho/work/ray/a.py", line 11, in <module>
    ray.get(f.remote())
  File "/Users/sangcho/work/ray/python/ray/remote_function.py", line 139, in _remote_proxy
    return self._remote(args=args, kwargs=kwargs, **self._default_options)
  File "/Users/sangcho/work/ray/python/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/Users/sangcho/work/ray/python/ray/util/tracing/tracing_helper.py", line 310, in _invocation_remote_span
    return method(self, args, kwargs, *_args, **_kwargs)
  File "/Users/sangcho/work/ray/python/ray/remote_function.py", line 304, in _remote
    self._pickled_function = pickle_dumps(
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 81, in pickle_dumps
    raise ray.exceptions.OufOfBandRefSerializationException(msg)
ray.exceptions.OufOfBandRefSerializationException: Could not serialize the function a.f:
=======================================================
Checking Serializability of <function f at 0x10bceb1f0>
=======================================================
!!! FAIL serialization: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1
Detected 1 global variables. Checking serializability...
    Serializing 'ref' ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505)...
    !!! FAIL serialization: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1
    WARNING: Did not find non-serializable object in ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505). This may be an oversight.
=======================================================
Variable: 

	FailTuple(ref [obj=ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505), parent=<function f at 0x10bceb1f0>])

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
=======================================================
Check https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting for more information.
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
=======================================================

```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/serialization.py [hunk 1]
```diff
@@ -208,13 +218,24 @@ def add_contained_object_ref(self, object_ref):
...
-            )
+            if not ALLOW_OUT_OF_BAND_OBJECT_REF_SERIALIZATION:
+                raise OufOfBandRefSerializationException(
+                    f"It is not allowed to serialize ray.ObjectRef {object_ref.hex()}."
+                    "If you want to allow serialization, "
+                    "set `RAY_allow_out_of_band_object_ref_serialization=1.` "
+                    "If you set the env var, the object is pinned forever in the "
+                    "lifetime of the worker process and can cause Ray object leaks."
+                    "See the trace to find where the serialization occurs: "
+                    f"{''.join(traceback.format_stack())}"
```
Comment (raulchen) [inline python/ray/_private/serialization.py:229]: just to confirm, we should be able to tell the containing object from this stack?

Comment (rkooo567) [reply python/ray/_private/serialization.py:229]: we can't. but we can see which line triggers the serialization. This is the error msg with your repro script.

```
ray::f() (pid=37282, ip=127.0.0.1)
  File "/Users/sangcho/work/ray/a.py", line 20, in f
    return ArrowBlockAccessor.numpy_to_block(batch)
  File "/Users/sangcho/work/ray/python/ray/data/_internal/arrow_block.py", line 249, in numpy_to_block
    col = ArrowPythonObjectArray.from_objects(col)
  File "/Users/sangcho/work/ray/python/ray/air/util/object_extensions/arrow.py", line 100, in from_objects
    dumped_bytes = pickle_dumps(
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
ray._private.serialization.OufOfBandRefSerializationException: It is not allowed to serialize ray.ObjectRef 00d950ec0ccf9d2affffffffffffffffffffffff0100000002e1f505.If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks.See the trace to find where the serialization occurs:   File "/Users/sangcho/work/ray/python/ray/_private/workers/default_worker.py", line 289, in <module>
    worker.main_loop()
  File "/Users/sangcho/work/ray/a.py", line 20, in f
    return ArrowBlockAccessor.numpy_to_block(batch)
  File "/Users/sangcho/work/ray/python/ray/data/_internal/arrow_block.py", line 249, in numpy_to_block
    col = ArrowPythonObjectArray.from_objects(col)
  File "/Users/sangcho/work/ray/python/ray/air/util/object_extensions/arrow.py", line 100, in from_objects
    dumped_bytes = pickle_dumps(
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
```

Comment (rkooo567) [reply python/ray/_private/serialization.py:229]: also new error with capturing object ref;

```
2024-09-06 17:47:28,806	INFO worker.py:1777 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
Traceback (most recent call last):
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 73, in pickle_dumps
    return pickle.dumps(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1479, in dumps
    cp.dump(obj)
  File "/Users/sangcho/work/ray/python/ray/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 152, in object_ref_reducer
    self.add_contained_object_ref(obj, obj.call_site())
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 221, in add_contained_object_ref
    raise ray.exceptions.OufOfBandRefSerializationException(
ray.exceptions.OufOfBandRefSerializationException: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sangcho/work/ray/a.py", line 11, in <module>
    ray.get(f.remote())
  File "/Users/sangcho/work/ray/python/ray/remote_function.py", line 139, in _remote_proxy
    return self._remote(args=args, kwargs=kwargs, **self._default_options)
  File "/Users/sangcho/work/ray/python/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/Users/sangcho/work/ray/python/ray/util/tracing/tracing_helper.py", line 310, in _invocation_remote_span
    return method(self, args, kwargs, *_args, **_kwargs)
  File "/Users/sangcho/work/ray/python/ray/remote_function.py", line 304, in _remote
    self._pickled_function = pickle_dumps(
  File "/Users/sangcho/work/ray/python/ray/_private/serialization.py", line 81, in pickle_dumps
    raise ray.exceptions.OufOfBandRefSerializationException(msg)
ray.exceptions.OufOfBandRefSerializationException: Could not serialize the function a.f:
=======================================================
Checking Serializability of <function f at 0x10bceb1f0>
=======================================================
!!! FAIL serialization: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1
Detected 1 global variables. Checking serializability...
    Serializing 'ref' ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505)...
    !!! FAIL serialization: It is not allowed to serialize ray.ObjectRef 00ffffffffffffffffffffffffffffffffffffff0100000001e1f505. If you want to allow serialization, set `RAY_allow_out_of_band_object_ref_serialization=1.` If you set the env var, the object is pinned forever in the lifetime of the worker process and can cause Ray object leaks. See the callsite and trace to find where the serialization occurs.
Callsite: Disabled. Set RAY_record_ref_creation_sites=1
    WARNING: Did not find non-serializable object in ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505). This may be an oversight.
=======================================================
Variable: 

	FailTuple(ref [obj=ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001e1f505), parent=<function f at 0x10bceb1f0>])

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
=======================================================
Check https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting for more information.
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
=======================================================

```

**Comment by rkooo567** [comment]
@angelinalg can you take a look at doc changes?

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+=================================================
+
+**TLDR:** Avoid serialize ``ray.ObjectRef`` because Ray cannot know when the GC the underlying object.
+
+Ray's ``ray.ObjectRef`` is distributed reference counted. Ray pins the underlying object until the reference is not used by the system anymore.
+When all references are gone, the pinned object is garbage collected and cleaned up from the system.
+However, if user code serializes ``ray.objectRef``, Ray cannot keep track of the reference.
+
+To avoid incorrect behavior, if ``ray.ObjectRef`` is serialized by ``ray.cloudpickle``, Ray pins the object for the lifetime of a worker. "pin" means that object cannot be evicted from the object store
+until the corresponding worker dies. It is prone to Ray object leaks, which can lead disk spilling.
```
Comment (jjyao) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:13]:
```suggestion
until the corresponding owner worker dies. It is prone to Ray object leaks, which can lead disk spilling.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 3]
```diff
@@ -0,0 +1,26 @@
+.. _ray-out-of-band-object-ref-serialization:
+
+Anti-pattern: Serialize ray.ObjectRef out of band
```
Comment (jjyao) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:3]: cross-reference it in https://docs.ray.io/en/latest/ray-core/objects/serialization.html#serializing-objectrefs

File: python/ray/exceptions.py [hunk 4]
```diff
@@ -829,6 +829,15 @@ class ObjectRefStreamEndOfStreamError(RayError):
     pass
 
 
+@DeveloperAPI
+class OufOfBandRefSerializationException(RayError):
```
Comment (jjyao) [inline python/ray/exceptions.py:833]:
```suggestion
class OufOfBandObjectRefSerializationException(RayError):
```

File: python/ray/exceptions.py [hunk 5]
```diff
@@ -829,6 +829,15 @@ class ObjectRefStreamEndOfStreamError(RayError):
...
+    """Raised when an `ray.ObjectRef` is out of band serialized by
+    `ray.cloudpickle`. It is an anti pattern.
```
Comment (jjyao) [inline python/ray/exceptions.py:835]: link to the doc?

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ray.init()
+
+
+@ray.remote
+def out_of_band_serialization_pickle():
+    obj_ref = ray.put(1)
+    import pickle
+
+    # object_ref is serialized from user code using a regular pickle.
+    # Ray cannot keep track of the reference, so the underlying object
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:16]:
```suggestion
    # Ray can't keep track of the reference, so the underlying object
```

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    # can be GC'ed unexpectedly which can cause unexpected hangs.
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:17]:
```suggestion
    # can be GC'ed unexpectedly, which can cause unexpected hangs.
```

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+
+@ray.remote
+def out_of_band_serialization_ray_cloudpickle():
+    obj_ref = ray.put(1)
+    from ray import cloudpickle
+
+    # ray.cloudpickle can serialize only when
+    # RAY_allow_out_of_band_object_ref_serialization=1 env var is set.
+    # However, the object_ref is pinned for the lifetime of the worker
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:28]:
```suggestion
    # However, the object_ref is pinned for the lifetime of the worker,
```

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+print("==== serialize object ref with pickle ====")
+result = ray.get(out_of_band_serialization_pickle.remote())
+try:
+    ray.get(pickle.loads(result), timeout=5)
+except ray.exceptions.GetTimeoutError:
+    print("Underlying object is unexpectedly GC'ed!\n\n")
+
+print("==== serialize object ref with ray.cloudpickle ====")
+# By default, it is not allowed to serialize ray.ObjectRef using
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:41]:
```suggestion
# By default, it's not allowed to serialize ray.ObjectRef using
```

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+try:
+    ray.get(out_of_band_serialization_ray_cloudpickle.remote())
+except Exception as e:
+    print(f"Exception raised from out_of_band_serialization_ray_cloudpickle {e}\n\n")
+
+print(
+    "==== serialize object ref with ray.cloudpickle with env var "
+    "RAY_allow_out_of_band_object_ref_serialization ===="
+)
+# It is allowed to use ray.cloudpickle to serialize object ref using an env var.
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:52]:
```suggestion
# It's allowed to use ray.cloudpickle to serialize object ref using an env var.
```

File: doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+ray.get(
+    out_of_band_serialization_ray_cloudpickle.options(
+        runtime_env={
+            "env_vars": {
+                "RAY_allow_out_of_band_object_ref_serialization": "1",
+            }
+        }
+    ).remote()
+)
+# you can see objects are stil pinned although it is GC'ed and not used anymore.
```
Comment (angelinalg) [inline doc/source/ray-core/doc_code/anti_pattern_out_of_band_object_ref_serialization.py:62]:
```suggestion
# you can see objects are stil pinned although it's GC'ed and not used anymore.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+=================================================
+
+**TLDR:** Avoid serialize ``ray.ObjectRef`` because Ray cannot know when the GC the underlying object.
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:6]:
```suggestion
**TLDR:** Avoid serializing ``ray.ObjectRef`` because Ray can't know when to garbage collect the underlying object.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+Ray's ``ray.ObjectRef`` is distributed reference counted. Ray pins the underlying object until the reference is not used by the system anymore.
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:8]:
```suggestion
Ray's ``ray.ObjectRef`` is distributed reference counted. Ray pins the underlying object until the reference isn't used by the system anymore.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+When all references are gone, the pinned object is garbage collected and cleaned up from the system.
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:9]:
```suggestion
When all references are gone, Ray garbage collects the pinned object and cleans it up from the system.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+However, if user code serializes ``ray.objectRef``, Ray cannot keep track of the reference.
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:10]:
```suggestion
However, if user code serializes ``ray.objectRef``, Ray can't keep track of the reference.
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+To avoid incorrect behavior, if ``ray.ObjectRef`` is serialized by ``ray.cloudpickle``, Ray pins the object for the lifetime of a worker. "pin" means that object cannot be evicted from the object store
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:12]:
```suggestion
To avoid incorrect behavior, if ``ray.cloudpickle`` serializes``ray.ObjectRef``, Ray pins the object for the lifetime of a worker. "Pin" means that object can't be evicted from the object store
```

File: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+until the corresponding worker dies. It is prone to Ray object leaks, which can lead disk spilling.
+
+To detect if this pattern exists in your code, you can set an environment variable ``RAY_allow_out_of_band_object_ref_serialization=0``. If Ray detects
+``ray.ObjectRef`` is serialized by ``ray.cloudpickle``, it raises an exception with helpful messages.
```
Comment (angelinalg) [inline doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst:16]:
```suggestion
that ``ray.cloudpickle`` serialized``ray.ObjectRef``, it raises an exception with helpful messages.
```

**Review by angelinalg — APPROVED** [review]
Sorry for the delay. Just some nits and suggestions to remove some passive voice. Please excuse any changes in meaning from my rewrites and correct as necessary.

File: python/ray/exceptions.py (hunk 5 above)
Comment (rkooo567) [reply python/ray/exceptions.py:835]: I can't yet (because doc is not there yet). we should do in a follow up

File: python/ray/_private/serialization.py [hunk 18]
```diff
@@ -48,11 +50,15 @@
...
     ObjectRefStreamEndOfStreamError,
 )
+import ray.exceptions
 from ray.experimental.compiled_dag_ref import CompiledDAGRef
 from ray.util import serialization_addons
 from ray.util import inspect_serializability
 
 logger = logging.getLogger(__name__)
+ALLOW_OUT_OF_BAND_OBJECT_REF_SERIALIZATION = bool(
+    int(os.getenv("RAY_allow_out_of_band_object_ref_serialization", "1"))
```
Comment (jjyao) [inline python/ray/_private/serialization.py:60]: Use `env_bool()`

File: python/ray/_private/serialization.py [hunk 19]
```diff
@@ -127,7 +136,7 @@ def actor_handle_reducer(obj):
             serialized, actor_handle_id, weak_ref = obj._serialization_helper()
             # Update ref counting for the actor handle
             if not weak_ref:
-                self.add_contained_object_ref(actor_handle_id)
+                self.add_contained_object_ref(actor_handle_id, True)
```
Comment (jjyao) [inline python/ray/_private/serialization.py:139]: why it's always True for this case

**Review by jjyao — APPROVED** [review]

File: python/ray/_private/serialization.py (hunk 19 above)
Comment (rkooo567) [reply python/ray/_private/serialization.py:139]: it crashes so many tests now. And I think the leak is probably very minimal for actor handle. I will add comments.

Comment (rkooo567) [reply python/ray/_private/serialization.py:139]: (it doesn't leak actual actors)

**Comment by rkooo567** [comment]
need approval from the data team. cc @raulchen

**Review by raulchen — APPROVED** [review]
